specs/distros.rst

```diff
@@ -0,0 +1,580 @@
+             as well as the description of :class:`Route` for the API of the route object.
+
+             Using ``route in network.routes()`` and ``network.routes().add(route)
+             removes the need for ``cloudbaseinit.osutils.check_static_route_exists``
+             and ``cloudbaseinit.osutils.add_static_route``.
+             """
+
+          def default_gateway(self):
+             """Get the default gateway.
+
+             Can be implemented in the terms of :meth:`routes`.
+             """
+
+         def interfaces(self):
+             """Get the network interfaces
+
+             This can be implemented in the same vein as :meth:`routes`, e.g.
+             ``InterfaceContainer([Interface(...), Interface(...), Interface(...)])``
+             """
+
+         def firewall_rules(self):
+             """Get a wrapper over the existing firewall rules.
+
+             Since this seems to be only used in Windows, it can be provided
+             only in the Windows utils.
+             The same behaviour as for :meth:`routes` can be used, that is:
+
+                 >>> rules = distro.network.firewall_rules()
+                 >>> rule = distro.network.FirewallRule(name=..., port=..., protocol=...)
+                 >>> rules.add(rule)
+                 >>> rules.delete(rule)
+                 >>> rule in rules
+                 >>> for rule in rules: print(rules)
+                 >>> del rules[i]
+                 >>> rule = rules[0]
+                 >>> rule.name, rule.port, rule.protocol, rule.allow
+
+             This gets rid of ``cloudbaseinit.osutils.firewall_add_rule`` and
+             ``cloudbaseinit.osutils.firewall_remove_rule``.
+             """
+
+         def set_static_network_config(self, adapter_name, address, netmask,
+                                       broadcast, gateway, dnsnameservers):
+             """Configure a new static network.
+
+             The :meth:``cloudinit.distros.Distro.apply_network`` should be
+             removed in the favour of this method,
+             which will be called by each network plugin.
+             The method can be a template method, providing
+             hooks for setting static DNS servers, setting static gateways or
+             setting static IP addresses, which will be implemented by specific
+             implementations of Distros.
+             """
+
+        def hosts(self):
+             """Get the content of /etc/hosts file in a more OO approach.
+
+
+             >>> hosts = distro.network.hosts()
+             >>> list(hosts) # support iteration and index access
+             >>> hosts.add(ipaddress, hostname, alias)
+             >>> hosts.delete(ipaddress, hostname, alias)
+
+             This gets rid of ``cloudinit.distros.Distro.update_etc_hosts``
+             and can provide support for adding a new hostname for Windows, as well.
+             """
+
+        class Route(object):
```

Comment:
Could be a namedtuple for start, indeed, but I don't want to leak this abstraction though (what we'll use at any point shouldn't matter, as long as we maintain the same API and behaviour of the object).